lib_neutral_prompt/cfg_denoiser_hijack.py

```diff
@@ -207,32 +208,162 @@ def get_salience(vector: torch.Tensor) -> torch.Tensor:
+    return vector * (vector.abs() >= top_k).to(vector.dtype)
 
 
-sd_samplers_hijacker = hijacker.ModuleHijacker.install_or_get(
-    module=sd_samplers,
-    hijacker_attribute='__neutral_prompt_hijacker',
-    on_uninstall=script_callbacks.on_script_unloaded,
-)
+try:
+    from ldm_patched.modules import samplers
+    from modules_forge import forge_sampler
+    forge = True
+except ImportError:
+    forge = False
 
 
-@sd_samplers_hijacker.hijack('create_sampler')
-def create_sampler_hijack(name: str, model, original_function):
-    sampler = original_function(name, model)
-    if not hasattr(sampler, 'model_wrap_cfg') or not hasattr(sampler.model_wrap_cfg, 'combine_denoised'):
-        if global_state.is_enabled:
-            warn_unsupported_sampler()
+if forge:
+    forge_sampler_hijacker = hijacker.ModuleHijacker.install_or_get(
+        module=forge_sampler,
+        hijacker_attribute='__forge_sampler_hijacker',
+        on_uninstall=script_callbacks.on_script_unloaded,
+    )
+    samplers_hijacker = hijacker.ModuleHijacker.install_or_get(
+        module=samplers,
+        hijacker_attribute='__samplers_hijacker',
+        on_uninstall=script_callbacks.on_script_unloaded,
+    )
 
-        return sampler
 
-    sampler.model_wrap_cfg.combine_denoised = functools.partial(
-        combine_denoised_hijack,
-        original_function=sampler.model_wrap_cfg.combine_denoised
+    @forge_sampler_hijacker.hijack('forge_sample')
+    def forge_sample(self, denoiser_params, cond_scale, cond_composition, original_function):
+        if not global_state.is_enabled:
+            return original_function(self, denoiser_params, cond_scale, cond_composition)
+
+        self.inner_model.inner_model.forge_objects.unet.model_options['cond_composition'] = cond_composition
+        self.inner_model.inner_model.forge_objects.unet.model_options['uncond'] = cond_from_a1111_to_patched_ldm(denoiser_params.text_uncond)
+
+        return original_function(self, denoiser_params, cond_scale, cond_composition)
+
+
+    def sampling_function_hijack(model, x, timestep, uncond, cond, cond_scale, model_options, seed, original_function):
+        if not global_state.is_enabled or not global_state.prompt_exprs:
+            return original_function(model, x, timestep, uncond, cond, cond_scale, model_options, seed)
+
+        prompt = global_state.prompt_exprs[0]
+        original_strengths, new_strengths = prompt.accept(ForgeStrengthTempOverride(), cond, 0)
+
+        model_options['original_strengths'] = original_strengths
+        return original_function(model, x, timestep, uncond, cond, cond_scale, model_options, seed)
+
+
+    class ForgeStrengthTempOverride:
+        def visit_leaf_prompt(
+            self,
+            that: neutral_prompt_parser.LeafPrompt,
+            cond: List[dict],
+            index: int,
+        ) -> tuple:
+            original_strength = cond[index].get('strength', 1.0)
+            new_strength = float(that.conciliation is None)
```

Comment:
could you explain the logic here? why `float(that.conciliation is None)` ?

Reply:
`that.conciliation is None` is true when the keyword for the prompt is `AND`. All other keywords are "auxiliary", they do not belong to the traditional composable diffusion process and should be ignored for the weighting. Does this help?

Reply:
If you know of a more readable way to structure the code, I'm all ears by the way.